### PR TITLE
Mesh deposition faster

### PR DESCRIPTION
**Pull request by cphyc**
## PR Summary

Similar to #1994 but with a faster implementation that make use of a cached field. I think the code could be made more robust / readable, so I'm keeping the WIP status.

## TODO
- [x] make code more readable
- [x] update doc to notify that the user has to manually access `sp[ptype, 'cell_index']` for faster access
- [x] make it work for patch-based data

### CODE REVIEW

**Comment by cphyc** [comment]
Compared to the version without cache, here are some results:
- access mesh field: 280ms
- access particle positions: 4.5s
- build index: 6.74s
- access deposited mesh field: ~5s
- access deposited mesh field (no index): 10s

**Comment by cphyc** [comment]
@yt-fido test this please

**Comment by cphyc** [comment]
@RicardaBeckmann I've updated the PR -- in particular I've changed the name of the operation to make it a bit more explicit and @ngoldbaum implemented the method for patch-based codes.

Your feedback would be valuable!

**Comment by RicardaBeckmann** [comment]
Good work Corentin, this is an awesome feature!

I've had a play around with it and I think there is an issue with how the center of objects is being set for  deposited mesh fields.

```python
sp=ds.sphere([0.5,0.5,0.5],ds.quan(1,'kpc'))
ds.add_mesh_sampling_particle_field(('index', 'radius'), ptype='gas_tracer')
fig,axes=plt.subplots(3,1,figsize=(8,6))

for ax,field in zip(axes,[('index','radius'),
                          ('gas_tracer','particle_radius'),
                          ('gas_tracer','cell_index_radius')]):
    data=sp[field].in_units('kpc')
    print(field,data.min(),data.max())
    ax.hist(data.v,label=field)
    ax.legend(loc=2)
    ax.set_xlabel('radius [kpc]')
    ax.set_ylabel('Number count')
```

produces

```
('index', 'radius') 0.10409343364780521 kpc 0.9966187511415499 kpc
('gas_tracer', 'particle_radius') 0.2619627299475482 kpc 0.9966187511410463 kpc
('gas_tracer', 'cell_index_radius') 6820.930435131786 kpc 6821.624383589753 kpc
```
and the following image
image


Clearly the radial values from the deposited field are wrong. I'm not quite sure where the error occurs, but maybe the center of the object isn't set correctly? For my current dataset 6821.6 kpc = 0.865 code_length so these values are way outside the confines of the original sphere

**Comment by cphyc** [comment]
That looks like a bug indeed. Did you already have the problem before my last modifications? Also do you have a similar issue when using another quantity (e.g. density, temperature)?

File: yt/geometry/oct_geometry_handler.py [hunk 1]
```diff
@@ -33,3 +36,83 @@ def get_smallest_dx(self):
...
+
+            # The index of the obj is stored on the first bits
+            Nobjs = len(data._current_chunk.objs)
+            Nbits = int(np.ceil(np.log2(Nobjs)))
+            icell = cell_index >> Nbits
+            iobj = cell_index - (icell << Nbits)
+            for i, subset in enumerate(data._current_chunk.objs):
+                mask = (iobj == i)
+
+                cell_data = subset[ftype, deposit_field].T.reshape(-1)
```
Comment (cphyc) [inline yt/geometry/oct_geometry_handler.py]: This line will return something unexpected for example if `deposit_field` is `('index', 'radius')`. The values returned by subset is given in absolute coordinates while it is expected to be in relative coordinates.

**Comment by cphyc** [comment]
With #c2d9c206be8496ed1092405e93fc2aab84a8e82d I get
```python
import yt
import matplotlib.pyplot as plt
import numpy as np

ds = yt.load('output_00080/info_00080.txt')

sp=ds.sphere([0.5,0.5,0.5], 0.3)
ds.add_mesh_sampling_particle_field(('index', 'radius'), ptype='io')
fig,axes=plt.subplots(3,1,figsize=(8,6))

bins = np.linspace(0, 20000, 40)
for ax,field in zip(axes,[('index','radius'),
                          ('io','particle_radius'),
                          ('io','cell_index_radius')]):
    data=sp[field].in_units('kpc')
    print(field,data.min(),data.max())
    ax.hist(data.v,label=field,bins=bins)
    ax.legend(loc=2)
    ax.set_xlabel('radius [kpc]')
    ax.set_ylabel('Number count')
```
image

**Comment by cphyc** [comment]
Are the test failures related?

**Comment by cphyc** [comment]
@yt-fido test this please

**Comment by cphyc** [comment]
I am confused about the py3 failures. Could anyone have a look at it? I'd say they are not related…

**Comment by cphyc** [comment]
@yt-fido test this please

**Comment by MatthewTurk** [comment]
This is on my queue to checkout and review -- awesome work so far.

**Comment by brittonsmith** [comment]
Also, I think the Python 3 failures are unrelated. The same thing is happening in at least one other PR. We'll have to look into that.

File: doc/source/analyzing/fields.rst [hunk 2]
```diff
@@ -501,6 +501,50 @@ available are:
...
+the field in the cell is assigned to the particle. This is for
+example useful when using tracer particles to have access to the
+Eulerian information for Lagrangian particles.
+
+The particle fields are named ``(ptype, cell_ftype_fname)`` where
+``ptype`` is the particle type onto which the deposition occurs,
+``ftype`` is the mesh field type (e.g. ``gas``) and ``fname`` is the
+field (e.g. ``temperature``, ``density``, ...). You can directly use
+the ``add_mesh_sampling_particle_field`` function defined on each dataset to
+depose a field onto the particles like so:
```
[the remark this answers is not in the document]
Comment (cphyc) [reply doc/source/analyzing/fields.rst]: That looks fine to me.

**Comment by cphyc** [comment]
@brittonsmith I have added a test using `ds.find_field_values_at_points` for the grid AMR type. However for octree datasets, I'm not sure how to do that.
Whenever I try to use `fake_amr_octree_ds`, I get the error `IOHandlerStreamOctree' object has no attribute '_read_particle_coords'`. Maybe I can use a RAMSES dataset instead?

**Comment by ngoldbaum** [comment]
I don't think particle I/O ever got implemented for the stream AMR octree handler.

**Comment by cphyc** [comment]
OK, in that case I'm using a small RAMSES dataset.

**Comment by matthewturk** [comment]
If these failures aren't real, I'm up for merging.

**Comment by Xarthisius** [comment]
@yt-fido test this please

**Comment by cphyc** [comment]
The failing tests look like they all come from the `PlotWindow` issue, so the tests failures seem unrelated.

**Comment by brittonsmith** [comment]
@yt-fido test this please

**Pull request merged**